src/wp-includes/formatting.php

```diff
@@ -1278,32 +1306,271 @@ function wp_check_invalid_utf8( $text, $strip = false ) {
+ * replacement characters in a row.
+ *
+ * Note! The Unicode Replacement Character is itself a Unicode character. It will not
+ * be possible to determine if its purpose in a string is intentional or if it comes
+ * as a result of replacing invalid bytes.
+ *
+ * Example:
+ *
+ *     // Valid strings come through unchanged.
+ *     'test' === wp_scrub_utf8( 'test' );
+ *
+ *     // Invalid sequences of bytes are replaced.
+ *     $invalid = "the byte \xC0 is never allowed in a UTF-8 string.";
+ *     "the byte \u{FFFD} is never allowed in a UTF-8 string." === wp_scrub_utf8( $invalid, true );
+ *     'the byte � is never allowed in a UTF-8 string.' === wp_scrub_utf8( $invalid, true );
+ *
+ *     // Maximal subparts are replaced individually.
+ *     '.�.' === wp_scrub_utf8( ".\xC0." );              // C0 is never valid.
+ *     '.�.' === wp_scrub_utf8( ".\xE2\x8C." );          // Missing A3 at end.
+ *     '.��.' === wp_scrub_utf8( ".\xE2\x8C\xE2\x8C." ); // Maximal subparts replaced separately.
+ *     '.��.' === wp_scrub_utf8( ".\xC1\xBF." );         // Overlong sequence.
+ *     '.���.' === wp_scrub_utf8( ".\xED\xA0\x80." );    // Surrogate half.
+ *
+ * @see https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-5/#G40630
+ *
+ * @since 6.9.0
+ *
+ * @param string $text String which is assumed to be UTF-8 but may contain invalid sequences of bytes.
+ * @return string Input text with invalid sequences of bytes replaced with the Unicode replacement character.
+ */
+function wp_scrub_utf8( $text ) {
+	static $use_fallback = null;
+
+	if ( null === $use_fallback ) {
+		$use_fallback = (
+			! extension_loaded( 'mbstring' ) ||
+			// Maximal subpart substitution introduced by php/php-src@04e59c916f12b322ac55f22314e31bd0176d01cb.
+			version_compare( PHP_VERSION, '8.1.6', '<' )
+		);
 	}
-	// We can't demand utf8 in the PCRE installation, so just return the string in those cases.
-	if ( ! $utf8_pcre ) {
-		return $text;
+
+	if ( $use_fallback ) {
+		return _wp_scrub_utf8_fallback( $text );
 	}
 
-	// phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged -- preg_match fails when it encounters invalid UTF8 in $text.
-	if ( 1 === @preg_match( '/^./us', $text ) ) {
-		return $text;
+	/*
+	 * While it looks like setting the substitute character could fail,
+	 * the internal PHP code will never fail when provided a valid
+	 * code point as a number. In this case, there’s no need to check
+	 * its return value to see if it succeeded.
+	 */
+	$prev_replacement_character = mb_substitute_character();
+	mb_substitute_character( 0xFFFD );
+	$scrubbed = mb_scrub( $text, 'UTF-8' );
+	mb_substitute_character( $prev_replacement_character );
+
+	/*
+	 * In PHP 8.0, `mb_scrub()` only returns a valid string. Once WordPress
+	 * depends on PHP 8.0.0 or above, this check will not be necessary.
+	 */
+	return false === $scrubbed
+		? _wp_scrub_utf8_fallback( $text )
+		: $scrubbed;
```

Comment:
Is this now covered by the `use_fallback` on PHP 8.1.6 check above? It seems like this should be unreachable.

Reply:
yes, very good catch. you can probably infer that I wrote this before adding the 8.1.6 check above.